# PyCity Schools Analysis

In [1]:
import pandas as pd
import numpy as np

In [2]:
df1 = pd.read_csv('schools_complete.csv')

In [3]:
df1.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [4]:
df2 = pd.read_csv('students_complete.csv')

In [5]:
df2.head()

Student ID       student_name gender grade        school_name   
0           0       Paul Bradley      M   9th  Huang High School  \
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [6]:
df1.shape

(15, 5)

In [7]:
df2.shape

(39170, 7)

In [8]:
df1.describe()

School ID         size        budget
count  15.000000    15.000000  1.500000e+01
mean    7.000000  2611.333333  1.643295e+06
std     4.472136  1420.915282  9.347763e+05
min     0.000000   427.000000  2.480870e+05
25%     3.500000  1698.000000  1.046265e+06
50%     7.000000  2283.000000  1.319574e+06
75%    10.500000  3474.000000  2.228999e+06
max    14.000000  4976.000000  3.124928e+06

In [9]:
df2.describe()

Student ID  reading_score    math_score
count  39170.000000    39170.00000  39170.000000
mean   19584.500000       81.87784     78.985371
std    11307.549359       10.23958     12.309968
min        0.000000       63.00000     55.000000
25%     9792.250000       73.00000     69.000000
50%    19584.500000       82.00000     79.000000
75%    29376.750000       91.00000     89.000000
max    39169.000000       99.00000     99.000000

In [10]:
school_data_complete = pd.merge(df2, df1, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID       student_name gender grade        school_name   
0           0       Paul Bradley      M   9th  Huang High School  \
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

# District Summary

In [11]:
#Calculate & Print District Summary 
unique_schools = np.unique(df1['school_name'])
total_unique_schools = len(unique_schools)
print("Total number of unique schools:", total_unique_schools)
total_students = df2.shape[0]
print("Total number of students:", total_students)
total_budget = np.sum(df1['budget'])
print("Total budget: $", total_budget)
average_math_score = np.mean(df2["math_score"])
print(f"Average math score: {average_math_score:.2f}%")
average_reading_score = np.mean(df2["reading_score"])
print(f"Average reading score: {average_reading_score:.2f}%")
passing_math_count = school_data_complete[(school_data_complete["math_score"] >= 70)].count()["student_name"]
passing_math_percentage = passing_math_count / float(total_students) * 100
passing_reading_count = school_data_complete[(school_data_complete["reading_score"] >= 70)].count()["student_name"]
passing_reading_percentage = passing_reading_count / float(total_students) * 100
passing_math_reading_count = school_data_complete[(school_data_complete["math_score"] >= 70) & (school_data_complete["reading_score"] >= 70)
].count()["student_name"]
overall_passing_rate = passing_math_reading_count / float(total_students) * 100
overall_passing_rate
print(f"Percentage of students passing math: {passing_math_percentage:.2f}%")
print(f"Percentage of students passing reading: {passing_reading_percentage:.2f}%")
print(f"Percentage of students passing both math and reading: {overall_passing_rate:.2f}%")

Total number of unique schools: 15
Total number of students: 39170
Total budget: $ 24649428
Average math score: 78.99%
Average reading score: 81.88%
Percentage of students passing math: 74.98%
Percentage of students passing reading: 85.81%
Percentage of students passing both math and reading: 65.17%


In [12]:
district_summary = pd.DataFrame({"Total Schools": [total_unique_schools], 
                                 "Total Students": [total_students], 
                                 "Total Budget": [total_budget],
                                 "Average Math Score": [average_math_score], 
                                 "Average Reading Score": [average_reading_score],
                                 "% Passing Math": [passing_math_percentage],
                                 "% Passing Reading": [passing_reading_percentage],
                                 "% Overall Passing Rate": [overall_passing_rate]})

district_summary = district_summary[["Total Schools", "Total Students", "Total Budget",
                                     "Average Math Score", 
                                     "Average Reading Score",
                                     "% Passing Math",
                                     "% Passing Reading",
                                     "% Overall Passing Rate"]]
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)
district_summary


Total Schools Total Students    Total Budget  Average Math Score   
0             15         39,170  $24,649,428.00           78.985371  \

   Average Reading Score  % Passing Math  % Passing Reading   
0               81.87784       74.980853          85.805463  \

   % Overall Passing Rate  
0               65.172326

In [13]:
#Calculate & Print School Summary 
    #Group data by school name 
gk = school_data_complete.groupby('school_name')
gk.first()

Student ID      student_name gender grade   
school_name                                                        
Bailey High School          17871      Blake Martin      M   9th  \
Cabrera High School         16013      Olivia Short      F  11th   
Figueroa High School         2917        Amy Jacobs      F  10th   
Ford High School            34796   Michael Mercado      M   9th   
Griffin High School         12262    Heather Wright      F  11th   
Hernandez High School        7627     Russell Davis      M  10th   
Holden High School          22847  Daniel Rodriguez      M  11th   
Huang High School               0      Paul Bradley      M   9th   
Johnson High School         30035        Lisa Casey      F  12th   
Pena High School            23274        Alec Davis      M   9th   
Rodriguez High School       26036    Sherry Jenkins      F  11th   
Shelton High School          5866  Jamie Montgomery      F  12th   
Thomas High School          37535        Norma Mata      F  10th   
Wilson High School          13730    Kelli Anderson      F  10th   
Wright High School          24236     Aaron Johnson      M  10th   

                       reading_score  math_score  School ID      type  size   
school_name                                                                   
Bailey High School                75          59          7  District  4976  \
Cabrera High School               94          94          6   Charter  1858   
Figueroa High School              85          87          1  District  2949   
Ford High School                  66          94         13  District  2739   
Griffin High School               79          68          4   Charter  1468   
Hernandez High School             70          88          3  District  4635   
Holden High School                86          92          8   Charter   427   
Huang High School                 66          79          0  District  2917   
Johnson High School               87          87         12  District  4761   
Pena High School                  91          75          9   Charter   962   
Rodriguez High School             74          81         11  District  3999   
Shelton High School               70          91          2   Charter  1761   
Thomas High School                76          76         14   Charter  1635   
Wilson High School                84          71          5   Charter  2283   
Wright High School                89          72         10   Charter  1800   

                        budget  
school_name                     
Bailey High School     3124928  
Cabrera High School    1081356  
Figueroa High School   1884411  
Ford High School       1763916  
Griffin High School     917500  
Hernandez High School  3022020  
Holden High School      248087  
Huang High School      1910635  
Johnson High School    3094650  
Pena High School        585858  
Rodriguez High School  2547363  
Shelton High School    1056600  
Thomas High School     1043130  
Wilson High School     1319574  
Wright High School     1049400

# Per School Summary

In [14]:
school_type = df1.set_index('school_name')['type']
total_students = school_data_complete.groupby('school_name')['Student ID'].count()
total_budget = school_data_complete.groupby('school_name')['budget'].mean()
per_student_budget = total_budget / total_students
average_math_score = school_data_complete.groupby('school_name')['math_score'].mean()
average_reading_score = school_data_complete.groupby('school_name')['reading_score'].mean()
passing_math = school_data_complete[school_data_complete['math_score'] >= 70].groupby('school_name')['Student ID'].count()
percent_passing_math = (passing_math / total_students) * 100
passing_reading = passing_reading = school_data_complete[school_data_complete['reading_score'] >= 70].groupby('school_name')['Student ID'].count()
percent_passing_reading = (passing_reading / total_students) * 100
passing_overall = school_data_complete[(school_data_complete['math_score'] >= 70) & (school_data_complete['reading_score'] >= 70)].groupby('school_name')['Student ID'].count()
percent_passing_overall = (passing_overall / total_students) * 100

In [15]:
per_school_summary = pd.concat([school_type,
                                total_students, 
                                total_budget,
                                per_student_budget,
                                average_math_score,
                                average_reading_score,
                                percent_passing_math.astype(float),
                                percent_passing_reading.astype(float),
                                percent_passing_overall.astype(float),
                               ],axis = 1)
per_school_summary.columns = ["School Type",
                              "Total Students",
                              "Total School Budget",
                              "Per Student Budget",
                              "Average Math Score",
                              "Average Reading Score",
                              "% Passing Math",
                              "% Passing Reading", 
                              "% Overall Passing",
                             ]

per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)

# Display the DataFrame
per_school_summary


School Type  Total Students Total School Budget   
school_name                                                             
Huang High School        District            2917       $1,910,635.00  \
Figueroa High School     District            2949       $1,884,411.00   
Shelton High School       Charter            1761       $1,056,600.00   
Hernandez High School    District            4635       $3,022,020.00   
Griffin High School       Charter            1468         $917,500.00   
Wilson High School        Charter            2283       $1,319,574.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Bailey High School       District            4976       $3,124,928.00   
Holden High School        Charter             427         $248,087.00   
Pena High School          Charter             962         $585,858.00   
Wright High School        Charter            1800       $1,049,400.00   
Rodriguez High School    District            3999       $2,547,363.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   
Thomas High School        Charter            1635       $1,043,130.00   

                      Per Student Budget  Average Math Score   
school_name                                                    
Huang High School                $655.00           76.629414  \
Figueroa High School             $639.00           76.711767   
Shelton High School              $600.00           83.359455   
Hernandez High School            $652.00           77.289752   
Griffin High School              $625.00           83.351499   
Wilson High School               $578.00           83.274201   
Cabrera High School              $582.00           83.061895   
Bailey High School               $628.00           77.048432   
Holden High School               $581.00           83.803279   
Pena High School                 $609.00           83.839917   
Wright High School               $583.00           83.682222   
Rodriguez High School            $637.00           76.842711   
Johnson High School              $650.00           77.072464   
Ford High School                 $644.00           77.102592   
Thomas High School               $638.00           83.418349   

                       Average Reading Score  % Passing Math   
school_name                                                    
Huang High School                  81.182722       65.683922  \
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Cabrera High School                83.975780       94.133477   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Rodriguez High School              80.744686       66.366592   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Thomas High School                 83.848930       93.272171   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Huang High School              81.316421          53.513884  
Figueroa High School           80.739234          53.204476  
Shelton High School            95.854628          89.892107  
Hernandez High School          80.862999          53.527508  
Griffin High School            97.138965          90.599455  
Wilson High School             96.539641          90.582567  
Cabrera High School            97.039828          91.334769  
Bailey High School             81.933280    

# Highest Performing Schools (by % Overall Passing)

In [16]:
top_schools = per_school_summary.sort_values("% Overall Passing", ascending = False)
top_schools.head(5)

School Type  Total Students Total School Budget   
school_name                                                           
Cabrera High School     Charter            1858       $1,081,356.00  \
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget  Average Math Score   
school_name                                                  
Cabrera High School            $582.00           83.061895  \
Thomas High School             $638.00           83.418349   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   
Pena High School               $609.00           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading   
school_name                                                                     
Cabrera High School              83.975780       94.133477          97.039828  \
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
school_name                             
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

# Bottom Performing Schools (by % Overall Passing)

In [17]:
bottom_schools = top_schools.tail()
bottom_schools = bottom_schools.sort_values('% Overall Passing')
bottom_schools.head(5)

School Type  Total Students Total School Budget   
school_name                                                             
Rodriguez High School    District            3999       $2,547,363.00  \
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Math Score   
school_name                                                    
Rodriguez High School            $637.00           76.842711  \
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Hernandez High School            $652.00           77.289752   
Johnson High School              $650.00           77.072464   

                       Average Reading Score  % Passing Math   
school_name                                                    
Rodriguez High School              80.744686       66.366592  \
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

# Math Scores by Grade

In [18]:

ninth_graders = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders = school_data_complete[(school_data_complete["grade"] == "12th")]

ninth_grader_math_scores = school_data_complete.loc[school_data_complete['grade'] == '9th'].groupby('school_name')['math_score'].mean()
tenth_grader_math_scores = school_data_complete.loc[school_data_complete['grade'] == '10th'].groupby('school_name')['math_score'].mean()
eleventh_grader_math_scores = school_data_complete.loc[school_data_complete['grade'] == '11th'].groupby('school_name')['math_score'].mean()
twelfth_grader_math_scores = school_data_complete.loc[school_data_complete['grade'] == '12th'].groupby('school_name')['math_score'].mean()

math_scores_by_grade = pd.DataFrame({
        "9th": ninth_grader_math_scores,
        "10th": tenth_grader_math_scores,
        "11th": eleventh_grader_math_scores,
        "12th": twelfth_grader_math_scores
})
math_scores_by_grade = math_scores_by_grade[['9th', '10th', '11th', '12th']]
math_scores_by_grade.index.name = None

math_scores_by_grade

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

# Reading Score By Grade

In [19]:
ninth_graders = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders = school_data_complete[(school_data_complete["grade"] == "12th")]

ninth_grader_reading_scores = school_data_complete.loc[school_data_complete['grade'] == '9th'].groupby('school_name')['reading_score'].mean()
tenth_grader_reading_scores = school_data_complete.loc[school_data_complete['grade'] == '10th'].groupby('school_name')['reading_score'].mean()
eleventh_grader_reading_scores = school_data_complete.loc[school_data_complete['grade'] == '11th'].groupby('school_name')['reading_score'].mean()
twelfth_grader_reading_scores = school_data_complete.loc[school_data_complete['grade'] == '12th'].groupby('school_name')['reading_score'].mean()

reading_scores_by_grade = pd.DataFrame({
        "9th": ninth_grader_reading_scores,
        "10th": tenth_grader_reading_scores,
        "11th": eleventh_grader_reading_scores,
        "12th": twelfth_grader_reading_scores
})

reading_scores_by_grade = reading_scores_by_grade[["9th", "10th", "11th", "12th"]]
reading_scores_by_grade.index.name = None

reading_scores_by_grade

9th       10th       11th       12th
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

# Scores by School Standing

In [20]:
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

school_spending_df = per_school_summary.copy()

school_spending_df["Spending Ranges (Per Student)"] = pd.cut(per_student_budget, spending_bins, labels=labels, right=False)
school_spending_df

School Type  Total Students Total School Budget   
school_name                                                             
Huang High School        District            2917       $1,910,635.00  \
Figueroa High School     District            2949       $1,884,411.00   
Shelton High School       Charter            1761       $1,056,600.00   
Hernandez High School    District            4635       $3,022,020.00   
Griffin High School       Charter            1468         $917,500.00   
Wilson High School        Charter            2283       $1,319,574.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Bailey High School       District            4976       $3,124,928.00   
Holden High School        Charter             427         $248,087.00   
Pena High School          Charter             962         $585,858.00   
Wright High School        Charter            1800       $1,049,400.00   
Rodriguez High School    District            3999       $2,547,363.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   
Thomas High School        Charter            1635       $1,043,130.00   

                      Per Student Budget  Average Math Score   
school_name                                                    
Huang High School                $655.00           76.629414  \
Figueroa High School             $639.00           76.711767   
Shelton High School              $600.00           83.359455   
Hernandez High School            $652.00           77.289752   
Griffin High School              $625.00           83.351499   
Wilson High School               $578.00           83.274201   
Cabrera High School              $582.00           83.061895   
Bailey High School               $628.00           77.048432   
Holden High School               $581.00           83.803279   
Pena High School                 $609.00           83.839917   
Wright High School               $583.00           83.682222   
Rodriguez High School            $637.00           76.842711   
Johnson High School              $650.00           77.072464   
Ford High School                 $644.00           77.102592   
Thomas High School               $638.00           83.418349   

                       Average Reading Score  % Passing Math   
school_name                                                    
Huang High School                  81.182722       65.683922  \
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Cabrera High School                83.975780       94.133477   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Rodriguez High School              80.744686       66.366592   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Thomas High School                 83.848930       93.272171   

                       % Passing Reading  % Overall Passing   
school_name                                                   
Huang High School              81.316421          53.513884  \
Figueroa High School           80.739234          53.204476   
Shelton High School            95.854628          89.892107   
Hernandez High School          80.862999          53.527508   
Griffin High School            97.138965          90.599455   
Wilson High School             96.539641          90.582567   
Cabrera High School            97.039828          91.334769   
Bailey High School             81.9

In [21]:
spending_math_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Math Score"].mean()
spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Reading Score"].mean()
spending_passing_math = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Math"].mean()
spending_passing_reading = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Reading"].mean()
overall_passing_spending = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Overall Passing"].mean()

spending_summary = pd.DataFrame({
    "Average Math Score": spending_math_scores,
    "Average Reading Score": spending_reading_scores,
    "% Passing Math": spending_passing_math,
    "% Passing Reading": spending_passing_reading,
    "% Overall Passing": overall_passing_spending            
})
spending_summary = spending_summary[[
    "Average Math Score",
    "Average Reading Score",
    "% Passing Math",
    "% Passing Reading",
    "% Overall Passing"
]]


spending_summary.index.name = "Per Student Budget"

spending_summary.style.format({'Average Math Score': '{:.2f}', 
                              'Average Reading Score': '{:.2f}', 
                              '% Passing Math': '{:.2f}', 
                              '% Passing Reading':'{:.2f}', 
                              '% Overall Passing': '{:.2f}'})
spending_summary

Average Math Score  Average Reading Score  % Passing Math   
Per Student Budget                                                              
<$585                        83.455399              83.933814       93.460096  \
$585-630                     81.899826              83.155286       87.133538   
$630-645                     78.518855              81.624473       73.484209   
$645-680                     76.997210              81.027843       66.164813   

                    % Passing Reading  % Overall Passing  
Per Student Budget                                        
<$585                       96.610877          90.369459  
$585-630                    92.718205          81.418596  
$630-645                    84.391793          62.857656  
$645-680                    81.133951          53.526855

# Scores By School Size

In [22]:
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], size_bins, labels=labels, right=False)
per_school_summary

School Type  Total Students Total School Budget   
school_name                                                             
Huang High School        District            2917       $1,910,635.00  \
Figueroa High School     District            2949       $1,884,411.00   
Shelton High School       Charter            1761       $1,056,600.00   
Hernandez High School    District            4635       $3,022,020.00   
Griffin High School       Charter            1468         $917,500.00   
Wilson High School        Charter            2283       $1,319,574.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Bailey High School       District            4976       $3,124,928.00   
Holden High School        Charter             427         $248,087.00   
Pena High School          Charter             962         $585,858.00   
Wright High School        Charter            1800       $1,049,400.00   
Rodriguez High School    District            3999       $2,547,363.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   
Thomas High School        Charter            1635       $1,043,130.00   

                      Per Student Budget  Average Math Score   
school_name                                                    
Huang High School                $655.00           76.629414  \
Figueroa High School             $639.00           76.711767   
Shelton High School              $600.00           83.359455   
Hernandez High School            $652.00           77.289752   
Griffin High School              $625.00           83.351499   
Wilson High School               $578.00           83.274201   
Cabrera High School              $582.00           83.061895   
Bailey High School               $628.00           77.048432   
Holden High School               $581.00           83.803279   
Pena High School                 $609.00           83.839917   
Wright High School               $583.00           83.682222   
Rodriguez High School            $637.00           76.842711   
Johnson High School              $650.00           77.072464   
Ford High School                 $644.00           77.102592   
Thomas High School               $638.00           83.418349   

                       Average Reading Score  % Passing Math   
school_name                                                    
Huang High School                  81.182722       65.683922  \
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Cabrera High School                83.975780       94.133477   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Rodriguez High School              80.744686       66.366592   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Thomas High School                 83.848930       93.272171   

                       % Passing Reading  % Overall Passing   
school_name                                                   
Huang High School              81.316421          53.513884  \
Figueroa High School           80.739234          53.204476   
Shelton High School            95.854628          89.892107   
Hernandez High School          80.862999          53.527508   
Griffin High School            97.138965          90.599455   
Wilson High School             96.539641          90.582567   
Cabrera High School            97.039828          91.334769   
Bailey High School             81.9

In [23]:
size_math_scores = per_school_summary.groupby(["School Size"])["Average Math Score"].mean()
size_reading_scores = per_school_summary.groupby(["School Size"])["Average Reading Score"].mean()
size_passing_math = per_school_summary.groupby(["School Size"])["% Passing Math"].mean()
size_passing_reading = per_school_summary.groupby(["School Size"])["% Passing Reading"].mean()
size_overall_passing = per_school_summary.groupby(["School Size"])["% Overall Passing"].mean()

size_summary = pd.DataFrame({
    "Average Math Score": size_math_scores,
    "Average Reading Score": size_reading_scores,
    "% Passing Math": size_passing_math,
    "% Passing Reading": size_passing_reading,
    "% Overall Passing": size_overall_passing            
})
size_summary = size_summary[[
    "Average Math Score",
    "Average Reading Score",
    "% Passing Math",
    "% Passing Reading",
    "% Overall Passing"
]]
size_summary.index.name = "School Size"

spending_summary.style.format({'Average Math Score': '{:.2f}', 
                              'Average Reading Score': '{:.2f}', 
                              '% Passing Math': '{:.2f}', 
                              '% Passing Reading':'{:.2f}', 
                              '% Overall Passing': '{:.2f}'})

size_summary


Average Math Score  Average Reading Score  % Passing Math   
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225  \
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

# Scores by School Type

In [24]:
school_type = school_data_complete.groupby("type")

average_math_score_by_type = per_school_summary.groupby(["School Type"])["Average Math Score"].mean()
average_reading_score_by_type = per_school_summary.groupby(["School Type"])["Average Reading Score"].mean()
average_percent_passing_math_by_type = per_school_summary.groupby(["School Type"])["% Passing Math"].mean()
average_percent_passing_reading_by_type = per_school_summary.groupby(["School Type"])["% Passing Reading"].mean()
average_percent_overall_passing_by_type = per_school_summary.groupby(["School Type"])["% Overall Passing"].mean()

type_summary = pd.DataFrame({
    "Average Math Score": average_math_score_by_type,
    "Average Reading Score": average_reading_score_by_type,
    '% Passing Math': average_percent_passing_math_by_type,
    '% Passing Reading': average_percent_passing_reading_by_type,
    "% Overall Passing": average_percent_overall_passing_by_type
})

type_summary = type_summary[[
    "Average Math Score",
    "Average Reading Score",
    '% Passing Math',
    '% Passing Reading',
    "% Overall Passing"
]]
type_summary.index.name = "School Type"

type_summary.style.format({'Average Math Score': '{:.6f}', 
                              'Average Reading Score': '{:.6f}', 
                              '% Passing Math': '{:.6f}', 
                              '% Passing Reading':'{:.6f}', 
                              '% Overall Passing': '{:.6f}'})

type_summary

Average Math Score  Average Reading Score  % Passing Math   
School Type                                                              
Charter               83.473852              83.896421       93.620830  \
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

### Based on the District Summary we can see that overall, the math and reading scores are = or >74% while the overall passing rate is much lower at 65%, indicating that the rate at which students are passing both classes is lower than the average scores. As Per School Summary, we can deduce that District school scores tend to be significantly lower as all District school scores sit around 55% for overall passing rate. Charter school scores, on the other hand, display significantly higher overall passing rates, mostly = or >90%. This is also proven by the Highest Performing Schools & Lowest Performing Schools dataframes, which display 5 Charter schools as top performers, and 5 District schools as bottom performers. While there is no clear relationship between per student/school budgets and overall performers, we can clearly see that school size and school type heavily impact student performance. We can conclude that the greater the school size, the lower the average & overall passing scores. Likewise, depending on whether the school type is District or Charter, the scores will be lower and greater, respectively.  